File: sdks/go/broker_test.go
package pilots

import (
	"os"
	"path/filepath"
	"testing"
	"time"
)

// A client inside a machine, with no key, uses the token the agent maintains.
func TestAClientWithNoKeyReadsTheMachinesToken(t *testing.T) {
	path := filepath.Join(t.TempDir(), "token")
	if err := os.WriteFile(path, []byte("pbt1.abc.def\n"), 0o600); err != nil {
		t.Fatalf("write: %v", err)
	}
	t.Setenv("PILOT_TOKEN_FILE", path)

	c := New("")
	if got := c.credential(); got != "pbt1.abc.def" {
		t.Errorf("credential = %q, want the token with its newline trimmed", got)
	}
}

// An explicit key is an explicit choice and must never be quietly replaced.
// Silently preferring a machine's own token would make a process that passed a
// key act as something else, which is the worst kind of authentication bug: the
// call succeeds, against the wrong identity.
func TestAnExplicitKeyIsNeverReplacedByTheMachinesToken(t *testing.T) {
	path := filepath.Join(t.TempDir(), "token")
	if err := os.WriteFile(path, []byte("pbt1.machine.token"), 0o600); err != nil {
		t.Fatalf("write: %v", err)
	}
	t.Setenv("PILOT_TOKEN_FILE", path)

	c := New("pilot_operator_key")
	if got := c.credential(); got != "pilot_operator_key" {
		t.Errorf("credential = %q, want the key that was passed", got)
	}
	if c.broker != nil {
		t.Error("a client given a key still built a broker credential")
	}
}

// No token file is the ordinary state of an ungranted machine. The client must
// send no credential rather than one saying "", and must not fail to construct.
func TestNoTokenFileIsNotAnError(t *testing.T) {
	t.Setenv("PILOT_TOKEN_FILE", filepath.Join(t.TempDir(), "absent"))
	c := New("")
	if got := c.credential(); got != "" {
		t.Errorf("credential = %q, want empty", got)
	}
}

// The token changes every five minutes, so a client must not hold one for the
// life of the process: that is a 401 with no cause anybody can see.
func TestAReplacedTokenIsPickedUp(t *testing.T) {
	path := filepath.Join(t.TempDir(), "token")
	if err := os.WriteFile(path, []byte("pbt1.first"), 0o600); err != nil {
		t.Fatalf("write: %v", err)
	}
	cred := &brokerCredential{path: path}
	if got := cred.Token(); got != "pbt1.first" {
		t.Fatalf("first read = %q", got)
	}

	if err := os.WriteFile(path, []byte("pbt1.second"), 0o600); err != nil {
		t.Fatalf("rewrite: %v", err)
	}
	// The cache is deliberately short rather than absent; aged out by hand here
	// so the test does not sleep for it.
	cred.readAt = time.Now().Add(-2 * brokerTokenTTL)
	if got := cred.Token(); got != "pbt1.second" {
		t.Errorf("second read = %q, want the replaced token", got)
	}
}

// Outside a machine there is no broker, and nothing should behave as though
// there were.
func TestInsideMachineIsFalseWithNoBroker(t *testing.T) {
	t.Setenv("PILOT_BROKER_URL", "")
	if InsideMachine() {
		t.Error("InsideMachine is true with no PILOT_BROKER_URL")
	}
	t.Setenv("PILOT_BROKER_URL", "http://169.254.0.22:3002")
	if !InsideMachine() {
		t.Error("InsideMachine is false with a broker address set")
	}
}

File: sdks/js/test/broker.test.ts
/**
 * A client inside a machine, with no key.
 */

import test from 'node:test'
import assert from 'node:assert/strict'
import { mkdtempSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

import { PilotsClient } from '../src/client.ts'
import { PilotsError } from '../src/errors.ts'

function tokenFile(contents: string): string {
  const path = join(mkdtempSync(join(tmpdir(), 'pilots-broker-')), 'token')
  writeFileSync(path, contents, { mode: 0o600 })
  return path
}

test('a client with no key uses the token the agent maintains', () => {
  const path = tokenFile('pbt1.abc.def\n')
  const previous = process.env.PILOT_TOKEN_FILE
  process.env.PILOT_TOKEN_FILE = path
  try {
    const client = new PilotsClient('', { baseURL: 'https://host-1.example.com' })
    assert.equal(client.apiKey, 'pbt1.abc.def', 'the trailing newline survived')
  } finally {
    if (previous === undefined) delete process.env.PILOT_TOKEN_FILE
    else process.env.PILOT_TOKEN_FILE = previous
  }
})

// An explicit key is an explicit choice. Silently preferring the machine's own
// token would make a process that passed a key act as something else, and the
// call would succeed against the wrong identity.
test('an explicit key is never replaced by the machine token', () => {
  const path = tokenFile('pbt1.machine.token')
  const previous = process.env.PILOT_TOKEN_FILE
  process.env.PILOT_TOKEN_FILE = path
  try {
    const client = new PilotsClient('pilot_operator_key', { baseURL: 'https://host-1.example.com' })
    assert.equal(client.apiKey, 'pilot_operator_key')
  } finally {
    if (previous === undefined) delete process.env.PILOT_TOKEN_FILE
    else process.env.PILOT_TOKEN_FILE = previous
  }
})

// Outside a machine, no key is still a mistake worth catching before any
// request: it would otherwise fail once per call with a 401 saying nothing.
test('no key and no machine token is still refused at construction', () => {
  const previous = process.env.PILOT_TOKEN_FILE
  delete process.env.PILOT_TOKEN_FILE
  try {
    assert.throws(() => new PilotsClient(''), PilotsError)
  } finally {
    if (previous !== undefined) process.env.PILOT_TOKEN_FILE = previous
  }
})
